pkg/apis/pipeline/v1beta1/merge.go

```diff
@@ -31,61 +31,111 @@ func MergeStepsWithStepTemplate(template *v1.Container, steps []Step) ([]Step, e
 		return steps, nil
 	}
 
-	// We need JSON bytes to generate a patch to merge the step containers
-	// onto the template container, so marshal the template.
-	templateAsJSON, err := json.Marshal(template)
```

Comment:
Probably a bit of an optimization but can we pass the marshall object in `mergeObject` instead of the template string ? Before this merge we would marshall the template once and the use it in the loop. With this change, we will marshall the template for each steps, even though it never changes.

Reply:
Done! Note that the template doesn't change for StepTemplates, but it does change for StepOverrides, so I've provided two separate functions